Adds support for specifying objectEncoding for secrets

chrisamert opened the pull request:
<!-- Thank you for helping Azure Key Vault Provider for Secrets Store CSI Driver with a pull request! Please make sure you read the [contributing guidelines](https://github.com/Azure/secrets-store-csi-driver-provider-azure#contributing). -->

**Reason for Change**:
Some secrets are stored encoded in KeyVault (for example, retrieving a pfx certificate from the secrets endpoint returns the base64 encoded certificate).  Many applications will be looking for or expecting the unencoded binary file.

This change adds support for specifying the objectEncoding of a secret.  The supported options are 'hex', 'base64' and 'utf-8'.  The default if not specified is 'utf-8' as that is the current behavior (value from KeyVault is written straight to a file).  If either 'hex' or 'base64' are specified, the secret string will first be decoded before being written to the file.


**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).


**Issue Fixed**:
Fixes #89 

**Please answer the following questions with yes/no**:

- Does this change contain code from or inspired by another project? If so, did you notify the maintainers and provide attribution?
No


**Special Notes for Reviewers**:
Docs and e2e tests will be coming, but I wanted to get this part out for review while I work on that

aramase commented on pkg/azure/provider.go:
```diff
@@ -669,3 +684,40 @@ func validateObjectFormat(objectFormat, objectType string) error {
+	}
+
+	if !strings.EqualFold(objectEncoding, objectEncodingHex) && !strings.EqualFold(objectEncoding, objectEncodingBase64) && !strings.EqualFold(objectEncoding, objectEncodingUtf8) {
+		return fmt.Errorf("Invalid objectEncoding: %v, should be Hex, Base64 or UTF-8", objectEncoding)
```
```suggestion
		return fmt.Errorf("Invalid objectEncoding: %v, should be hex, base64 or utf-8", objectEncoding)
```

aramase commented on pkg/azure/provider.go:
```diff
@@ -669,3 +684,40 @@ func validateObjectFormat(objectFormat, objectType string) error {
+		return hex.DecodeString(content)
+	}
+
+	return make([]byte, 0), fmt.Errorf("Invalid objectEncoding. Should be UTF-8, base64, or hex")
```
```suggestion
	return make([]byte, 0), fmt.Errorf("Invalid objectEncoding. Should be utf-8, base64, or hex")
```

aramase reviewed (commented):
@chrisamert Thank you for the PR 🎉 

1. Could we also add an e2e test here - https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/test/bats/azure.bats#L193-L195. Those 3 lines get the raw PFX data, base64 decode and then convert with openssl. For this PR we could add another snippet with object encoding [here](https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/test/bats/tests/certificates/azure_v1alpha1_secretproviderclass.yaml#L62-L67) and then add a test to validate with openssl without manually base64 decoding.
2. We also need to add the new fields to the table here - https://github.com/Azure/secrets-store-csi-driver-provider-azure#create-your-own-secretproviderclass-object

Please feel free to reach out if you have any questions.

chrisamert commented:
@aramase thanks for the review, I've added the e2e tests and updated the readme.

Prior to rebasing this change ontop of the latest changes I had the e2e tests passing locally.  However after rebasing (due to merge conflicts) I'm hitting some errors with the e2e tests.  Specifically the 'csi-csi-secrets-store-provider-azure' pod is crashing on startup with the message:

```
unknown flag: --endpoint
Usage of /bin/secrets-store-csi-driver-provider-azure:
      --attributes string     volume attributes
      --context-timeout int   context timeout in seconds for provider calls (default 110)
      --debug                 sets log to debug level
      --permission string     File permission
      --secrets string        node publish ref secret
      --targetPath string     Target path to write data.
      --version               prints the version information
```

This doesn't look directly related to my changes, but I'm out of time investigating today.  If you know off the top of your head what the issue could be here let me know!  Otherwise I'll look into it more tomorrow.

chrisamert commented:
> @aramase thanks for the review, I've added the e2e tests and updated the readme.
> 
> Prior to rebasing this change ontop of the latest changes I had the e2e tests passing locally. However after rebasing (due to merge conflicts) I'm hitting some errors with the e2e tests. Specifically the 'csi-csi-secrets-store-provider-azure' pod is crashing on startup with the message:
> 
> ```
> unknown flag: --endpoint
> Usage of /bin/secrets-store-csi-driver-provider-azure:
>       --attributes string     volume attributes
>       --context-timeout int   context timeout in seconds for provider calls (default 110)
>       --debug                 sets log to debug level
>       --permission string     File permission
>       --secrets string        node publish ref secret
>       --targetPath string     Target path to write data.
>       --version               prints the version information
> ```
> 
> This doesn't look directly related to my changes, but I'm out of time investigating today. If you know off the top of your head what the issue could be here let me know! Otherwise I'll look into it more tomorrow.

Ignore this, I got it figured out.  Didn't do a full build after the rebase :)

aramase commented on test/bats/azure.bats:
```diff
@@ -175,6 +175,11 @@ setup() {
   diff  <(echo "${result_base64_encoded}" ) <(echo "${CERT2_SECRET_VALUE}")
   [[ "${result_base64_encoded}" == "${CERT2_SECRET_VALUE}" ]]
+
+  result=$(kubectl exec nginx-secrets-store-inline-crd-certs -- $EXEC_COMMAND/$CERT2_NAME-secret-pfx-binary | base64)
```
nit: 

We can skip the `base64 -d` in `result_base64_encoded` instead of encoding it and then decoding.  

```bash
  result=$(kubectl exec nginx-secrets-store-inline-crd-certs -- $EXEC_COMMAND/$CERT2_NAME-secret-pfx-binary)
  result_base64_encoded=$(echo "${result//$'\r'}" | openssl pkcs12 -nodes -passin pass:"" | sed -ne '/-BEGIN PRIVATE KEY-/,/-END PRIVATE KEY-/p; /-BEGIN CERTIFICATE-/,/-END CERTIFICATE-/p' | base64 ${BASE64_FLAGS})
  [[ "${result_base64_encoded}" == "${CERT2_SECRET_VALUE}" ]]

```

chrisamert replied:
I did try that first, but running just 'cat <binary file>' in kubectl exec corrupted the binary data, and the next call to openssl failed with: asn1 encoding routines:asn1_d2i_read_bio:not enough data

Encoding the file inside the exec statement was enough to get around this so I went with that.  If we want to do this without the base64 encoding I think we'd want to use 'kubectl cp' instead of 'exec'.  That adds a bit of complexity around local file handling in the azure.bats file, but it's something I could look into

aramase replied:
Thank you for the explanation. That makes sense. We can keep it the way it is now.

aramase reviewed (commented):
@chrisamert Just one very small nit. Once that's updated we can merge the PR!

aramase commented on pkg/provider/provider.go:
```diff
@@ -520,3 +534,40 @@ func getObjectVersion(id string) string {
+
+	// ObjectEncoding is supported only for secret types
+	if objectType != VaultObjectTypeSecret {
+		return fmt.Errorf("ObjectEncoding only supported for objectType: secret")
```
nit: Error string should be capitalized 

```suggestion
		return fmt.Errorf("objectEncoding only supported for objectType: secret")
```

aramase commented on pkg/provider/provider.go:
```diff
@@ -520,3 +534,40 @@ func getObjectVersion(id string) string {
+	}
+
+	if !strings.EqualFold(objectEncoding, objectEncodingHex) && !strings.EqualFold(objectEncoding, objectEncodingBase64) && !strings.EqualFold(objectEncoding, objectEncodingUtf8) {
+		return fmt.Errorf("Invalid objectEncoding: %v, should be hex, base64 or utf-8", objectEncoding)
```
nit: Error string should be capitalized

```suggestion
		return fmt.Errorf("invalid objectEncoding: %v, should be hex, base64 or utf-8", objectEncoding)
```

aramase commented on pkg/provider/provider.go:
```diff
@@ -520,3 +534,40 @@ func getObjectVersion(id string) string {
+	return nil
+}
+
+// getObjectBytes takes the given content string and returns the bytes to write to disk
```
```suggestion
// getContentBytes takes the given content string and returns the bytes to write to disk
```

aramase commented on pkg/provider/provider.go:
```diff
@@ -520,3 +534,40 @@ func getObjectVersion(id string) string {
+		return hex.DecodeString(content)
+	}
+
+	return make([]byte, 0), fmt.Errorf("Invalid objectEncoding. Should be utf-8, base64, or hex")
```
```suggestion
	return make([]byte, 0), fmt.Errorf("invalid objectEncoding. Should be utf-8, base64, or hex")
```

aramase commented on test/bats/azure.bats:
```diff
@@ -175,6 +175,11 @@ setup() {
   diff  <(echo "${result_base64_encoded}" ) <(echo "${CERT2_SECRET_VALUE}")
   [[ "${result_base64_encoded}" == "${CERT2_SECRET_VALUE}" ]]
+
+  result=$(kubectl exec nginx-secrets-store-inline-crd-certs -- $EXEC_COMMAND/$CERT2_NAME-secret-pfx-binary | base64)
```
```suggestion
  result=$(kubectl exec nginx-secrets-store-inline-crd-certs -- $EXEC_COMMAND/$CERT2_NAME-secret-pfx-binary | base64 ${BASE64_FLAGS})
```

aramase commented:
/azp run pr-e2e-azure

aramase commented:
/azp run pr-e2e-azure

aramase reviewed (approved):
lgtm

ritazh reviewed (approved):
lgtm
